In [1]:
#Importing essential libraries

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
#passing the csv file into pd.read_csv() to create a DataFrame
quiz_data = pd.read_csv('energydata_complete.csv')

In [3]:
quiz_data

date  Appliances  lights         T1       RH_1  \
0      2016-01-11 17:00:00          60      30  19.890000  47.596667   
1      2016-01-11 17:10:00          60      30  19.890000  46.693333   
2      2016-01-11 17:20:00          50      30  19.890000  46.300000   
3      2016-01-11 17:30:00          50      40  19.890000  46.066667   
4      2016-01-11 17:40:00          60      40  19.890000  46.333333   
...                    ...         ...     ...        ...        ...   
19730  2016-05-27 17:20:00         100       0  25.566667  46.560000   
19731  2016-05-27 17:30:00          90       0  25.500000  46.500000   
19732  2016-05-27 17:40:00         270      10  25.500000  46.596667   
19733  2016-05-27 17:50:00         420      10  25.500000  46.990000   
19734  2016-05-27 18:00:00         430      10  25.500000  46.600000   

              T2       RH_2         T3       RH_3         T4  ...         T9  \
0      19.200000  44.790000  19.790000  44.730000  19.000000  ...  17.033333   
1      19.200000  44.722500  19.790000  44.790000  19.000000  ...  17.066667   
2      19.200000  44.626667  19.790000  44.933333  18.926667  ...  17.000000   
3      19.200000  44.590000  19.790000  45.000000  18.890000  ...  17.000000   
4      19.200000  44.530000  19.790000  45.000000  18.890000  ...  17.000000   
...          ...        ...        ...        ...        ...  ...        ...   
19730  25.890000  42.025714  27.200000  41.163333  24.700000  ...  23.200000   
19731  25.754000  42.080000  27.133333  41.223333  24.700000  ...  23.200000   
19732  25.628571  42.768571  27.050000  41.690000  24.700000  ...  23.200000   
19733  25.414000  43.036000  26.890000  41.290000  24.700000  ...  23.200000   
19734  25.264286  42.971429  26.823333  41.156667  24.700000  ...  23.200000   

          RH_9      T_out  Press_mm_hg     RH_out  Windspeed  Visibility  \
0      45.5300   6.600000        733.5  92.000000   7.000000   63.000000   
1      45.5600   6.483333        733.6  92.000000   6.666667   59.166667   
2      45.5000   6.366667        733.7  92.000000   6.333333   55.333333   
3      45.4000   6.250000        733.8  92.000000   6.000000   51.500000   
4      45.4000   6.133333        733.9  92.000000   5.666667   47.666667   
...        ...        ...          ...        ...        ...         ...   
19730  46.7900  22.733333        755.2  55.666667   3.333333   23.666667   
19731  46.7900  22.600000        755.2  56.000000   3.500000   24.500000   
19732  46.7900  22.466667        755.2  56.333333   3.666667   25.333333   
19733  46.8175  22.333333        755.2  56.666667   3.833333   26.166667   
19734  46.8450  22.200000        755.2  57.000000   4.000000   27.000000   

       Tdewpoint        rv1        rv2  
0       5.300000  13.275433  13.275433  
1       5.200000  18.606195  18.606195  
2       5.100000  28.642668  28.642668  
3       5.000000  45.410389  45.410389  
4       4.900000  10.084097  10.084097  
...          ...        ...        ...  
19730  13.333333  43.096812  43.096812  
19731  13.300000  49.282940  49.282940  
19732  13.266667  29.199117  29.199117  
19733  13.233333   6.322784   6.322784  
19734  13.200000  34.118851  34.118851  

[19735 rows x 29 columns]

In [5]:
#Normalising the quiz data

from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()

quiz_df = quiz_data.drop(['date', 'lights'], axis =1)

normalised_quiz_df = pd.DataFrame(scaler.fit_transform(quiz_df), columns = quiz_df.columns)
normalised_quiz_df

Appliances        T1      RH_1        T2      RH_2        T3      RH_3  \
0        0.046729  0.327350  0.566187  0.225345  0.684038  0.215188  0.746066   
1        0.046729  0.327350  0.541326  0.225345  0.682140  0.215188  0.748871   
2        0.037383  0.327350  0.530502  0.225345  0.679445  0.215188  0.755569   
3        0.037383  0.327350  0.524080  0.225345  0.678414  0.215188  0.758685   
4        0.046729  0.327350  0.531419  0.225345  0.676727  0.215188  0.758685   
...           ...       ...       ...       ...       ...       ...       ...   
19730    0.084112  0.926786  0.537657  0.711655  0.606309  0.830841  0.579374   
19731    0.074766  0.919747  0.536006  0.701769  0.607836  0.825302  0.582178   
19732    0.242991  0.919747  0.538666  0.692651  0.627198  0.818378  0.603988   
19733    0.383178  0.919747  0.549491  0.677054  0.634717  0.805085  0.585294   
19734    0.392523  0.919747  0.538758  0.666171  0.632902  0.799546  0.579062   

             T4      RH_4        T5  ...        T9      RH_9     T_out  \
0      0.351351  0.764262  0.175506  ...  0.223032  0.677290  0.372990   
1      0.351351  0.782437  0.175506  ...  0.226500  0.678532  0.369239   
2      0.344745  0.778062  0.175506  ...  0.219563  0.676049  0.365488   
3      0.341441  0.770949  0.175506  ...  0.219563  0.671909  0.361736   
4      0.341441  0.762697  0.178691  ...  0.219563  0.671909  0.357985   
...         ...       ...       ...  ...       ...       ...       ...   
19730  0.864865  0.765258  0.752031  ...  0.864724  0.729443  0.891747   
19731  0.864865  0.765258  0.754897  ...  0.864724  0.729443  0.887460   
19732  0.864865  0.771233  0.754897  ...  0.864724  0.729443  0.883173   
19733  0.864865  0.773794  0.752031  ...  0.864724  0.730581  0.878885   
19734  0.864865  0.781192  0.752031  ...  0.864724  0.731719  0.874598   

       Press_mm_hg    RH_out  Windspeed  Visibility  Tdewpoint       rv1  \
0         0.097674  0.894737   0.500000    0.953846   0.538462  0.265449   
1         0.100000  0.894737   0.476190    0.894872   0.533937  0.372083   
2         0.102326  0.894737   0.452381    0.835897   0.529412  0.572848   
3         0.104651  0.894737   0.428571    0.776923   0.524887  0.908261   
4         0.106977  0.894737   0.404762    0.717949   0.520362  0.201611   
...            ...       ...        ...         ...        ...       ...   
19730     0.602326  0.416667   0.238095    0.348718   0.901961  0.861981   
19731     0.602326  0.421053   0.250000    0.361538   0.900452  0.985726   
19732     0.602326  0.425439   0.261905    0.374359   0.898944  0.583979   
19733     0.602326  0.429825   0.273810    0.387179   0.897436  0.126371   
19734     0.602326  0.434211   0.285714    0.400000   0.895928  0.682391   

            rv2  
0      0.265449  
1      0.372083  
2      0.572848  
3      0.908261  
4      0.201611  
...         ...  
19730  0.861981  
19731  0.985726  
19732  0.583979  
19733  0.126371  
19734  0.682391  

[19735 rows x 27 columns]

In [7]:
# Import train_test_split to split the dataset into train and test
from sklearn.model_selection import train_test_split
X = normalised_quiz_df.drop('Appliances', axis = 1)
y = normalised_quiz_df['Appliances']
X_train, X_valid, y_train, y_valid = train_test_split(X, y, test_size=0.3, random_state=42)

In [37]:
#Linear Regression between T2 and T6
from sklearn.linear_model import LinearRegression

linear_model= LinearRegression()

linear_model.fit(X_train[['T2']], X_train[['T6']])

round(linear_model.score(X_valid[['T2']], X_valid[['T6']]), 2)

0.64

The $r^2$ value of T2 and T6 is **0.64**

In [38]:
linear = LinearRegression()

linear.fit(X_train, y_train)
predict= linear.predict(X_valid)
round(mean_absolute_error(y_valid, predict), 2)

0.05

The **MEAN ABSOLUTE ERROR** is 0.05

In [39]:
#defing a function to calculate the residual sum of squares
def RSS(actual, predicted):
    return np.sum(np.square(actual - predicted))
round(RSS(y_valid, predict), 2)

45.35

In [46]:
round(np.sqrt(mean_squared_error(y_valid, predict)), 3)

0.088

##### root mean squared value for the linear_model is  0.088

In [41]:
round(linear.score(X_valid, y_valid), 2)

0.15

In [44]:
data = pd.DataFrame()
data['weights'] = linear.coef_
data['columns'] = X_train.columns
data.sort_values('weights').reset_index(drop=True).iloc[[0, -1], :]

weights columns
0  -0.456698    RH_2
25  0.553547    RH_1

The features with the highest and lowest weights are **RH_2** and **RH_1**

In [45]:
# Import Ridge to check for the 

from sklearn.linear_model import Ridge, Lasso

ridge = Ridge(alpha=0.4)
ridge.fit(X_train, y_train)
pred_ridge = ridge.predict(X_valid)
round(np.sqrt(mean_squared_error(y_valid, pred_ridge)), 3)

0.088

##### Root mean squared value for linear_model and Ridge are the same

In [33]:
lasso = Lasso(alpha=0.001)

lasso.fit(X_train, y_train)

datlas = pd.DataFrame()
datlas['weights'] = lasso.coef_
datlas['columns'] = X_train.columns
datlas.sort_values('weights')

weights      columns
20 -0.049557       RH_out
15 -0.000110         RH_8
0   0.000000           T1
23  0.000000    Tdewpoint
22  0.000000   Visibility
19 -0.000000  Press_mm_hg
18  0.000000        T_out
17 -0.000000         RH_9
16 -0.000000           T9
14  0.000000           T8
13 -0.000000         RH_7
24 -0.000000          rv1
12 -0.000000           T7
10  0.000000           T6
9   0.000000         RH_5
8  -0.000000           T5
7   0.000000         RH_4
6  -0.000000           T4
5   0.000000         RH_3
4   0.000000           T3
3  -0.000000         RH_2
2   0.000000           T2
11 -0.000000         RH_6
25 -0.000000          rv2
21  0.002912    Windspeed
1   0.017880         RH_1

In [34]:
datlas[datlas['weights'] != 0]

weights    columns
1   0.017880       RH_1
15 -0.000110       RH_8
20 -0.049557     RH_out
21  0.002912  Windspeed

The Total Number of features with Non Zeros using **Lasso** are **4**

In [48]:
pred_lasso = lasso.predict(X_valid)
round(np.sqrt(mean_squared_error(y_valid, pred_lasso)), 3)

0.094

##### Root Mean Squared value for Lasso model is 0.094